File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.go
func maxUniqueSplit(s string) int {
    var uniq = make(map[string]bool)
    var splitsCounts func(start, end int) int;
    
    splitsCounts = func (start, end int) int {
        if end >= len(s) {
            return 0
        }
        
        var ret = splitsCounts(start, end + 1)
        if !uniq[s[start : end + 1]] {
            uniq[s[start : end + 1]] = true
            
            ret = max(ret, 1 + splitsCounts(end + 1, end + 1))
            uniq[s[start : end + 1]] = false
        }
        return ret
    }
    
    return splitsCounts(0, 0)
}

File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        
        def max_splits(idx, seen = set()):
            nonlocal s
            if idx >= len(s):
                return len(seen)
            
            ret = 0
            for i in range(idx, len(s)):
                if s[idx: i + 1] not in seen:
                    ret = max(ret, max_splits(i + 1, seen | {s[idx: i + 1]}))
            
            
            return ret
        
        
        return max_splits(0)
                    
                